File: test-validation/python/xxe/vulnerable.py
"""
漏洞示例：XXE（XML External Entity）注入 - Python

问题说明：
使用 lxml 的 etree.parse() 解析 XML 时，未禁用外部实体解析，
攻击者可以构造恶意 XML 读取服务器文件或发起 SSRF 攻击。

预期检出：
- 行号：22
- 规则 ID：xxe-python-lxml-parse
- 严重级别：ERROR
"""

from lxml import etree


def parse_xml_unsafe(xml_string: str) -> etree._Element:
    """漏洞：使用 lxml 解析 XML 时未禁用外部实体"""
    # 漏洞：直接使用默认解析器，未禁用外部实体
    parser = etree.XMLParser()  # 第 21 行 - 不安全的解析器
    root = etree.fromstring(xml_string.encode(), parser)  # 第 22 行 - XXE 漏洞点
    return root


def parse_xml_file_unsafe(file_path: str) -> etree._ElementTree:
    """漏洞：解析 XML 文件时未禁用外部实体"""
    # 漏洞：直接使用 etree.parse，未设置安全解析器
    tree = etree.parse(file_path)  # 第 28 行 - XXE 漏洞点
    return tree


def parse_xml_with_resolve(xml_string: str) -> etree._Element:
    """漏洞：显式启用 resolve_entities"""
    # 漏洞：显式设置 resolve_entities=True
    parser = etree.XMLParser(resolve_entities=True)  # 第 34 行 - XXE 漏洞点
    root = etree.fromstring(xml_string.encode(), parser)
    return root

File: test-validation/java/xxe/Vulnerable.java
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.DocumentBuilder;
import org.w3c.dom.Document;
import java.io.InputStream;

/**
 * 漏洞示例：XXE（XML External Entity）注入
 *
 * 问题说明：
 * DocumentBuilderFactory 创建后未禁用外部实体解析，
 * 攻击者可以构造恶意 XML 读取服务器文件或发起 SSRF 攻击。
 *
 * 预期检出：
 * - 行号：16
 * - 规则 ID：xxe-java-document-builder
 * - 严重级别：ERROR
 */
public class Vulnerable {

    public Document parseXml(InputStream input) throws Exception {
        // 漏洞：未禁用外部实体，直接创建解析器
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        return builder.parse(input);  // 第 23 行 - XXE 漏洞点
    }

    public Document parseXmlFromFile(String filePath) throws Exception {
        // 漏洞：同样未禁用外部实体
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        return builder.parse(new java.io.File(filePath));  // 第 29 行 - XXE 漏洞点
    }
}
